client/ipa-join.c

```diff
@@ -612,161 +632,427 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
     return rval;
 }
 
+#else
+
+static inline struct curl_slist *
+curl_slist_append_log(struct curl_slist *list, char *string, int quiet) {
+    list = curl_slist_append(list, string);
+    if (!list) {
+        if (!quiet)
+            fprintf(stderr, _("curl_slist_append() failed for value: '%s'\n"), string);
+        return NULL;
+    }
+    return list;
+}
+
+#define CURL_SETOPT(curl, opt, val) \
+    if (curl_easy_setopt(curl, opt, val) != CURLE_OK) { \
+        if (!quiet) \
+            fprintf(stderr, _("curl_easy_setopt() failed\n")); \
+        rval = 17; \
+        goto cleanup; \
+    }
+
+size_t
+jsonrpc_handle_response(char *ptr, size_t size, size_t nmemb, void *userdata) {
+    size_t realsize = size * nmemb;
+    curl_buffer *cb = (curl_buffer *) userdata;
+
+    char *buf = (char *) realloc(cb->payload, cb->size + realsize + 1);
+    if (!buf) {
+        fprintf(stderr, _("Expanding buffer in jsonrpc_handle_response failed"));
+        free(buf);
+        cb->payload = NULL;
+        return 0;
+    }
+    cb->payload = buf;
+    memcpy(&(cb->payload[cb->size]), ptr, realsize);
+
+    cb->size += realsize;
+    cb->payload[cb->size] = 0;
+
+    return realsize;
+}
+
 static int
-unenroll_host(const char *server, const char *hostname, const char *ktname, int quiet)
-{
+jsonrpc_request(const char *ipaserver, const json_t *json, curl_buffer *response, bool quiet) {
     int rval = 0;
-    int ret;
-    char *ipaserver = NULL;
-    char *host = NULL;
-    struct utsname uinfo;
-    char *principal = NULL;
-    char *realm = NULL;
 
-    krb5_context krbctx = NULL;
-    krb5_keytab keytab = NULL;
-    krb5_ccache ccache = NULL;
-    krb5_principal princ = NULL;
-    krb5_error_code krberr;
-    krb5_creds creds;
-    krb5_get_init_creds_opt gicopts;
-    char tgs[LINE_MAX];
+    CURL *curl = NULL;
 
-    xmlrpc_env env;
-    xmlrpc_value * argArrayP = NULL;
-    xmlrpc_value * paramArrayP = NULL;
-    xmlrpc_value * paramP = NULL;
-    xmlrpc_value * resultP = NULL;
-    xmlrpc_server_info * serverInfoP = NULL;
-    xmlrpc_value *princP = NULL;
-    char * url = NULL;
-    char * user_agent = NULL;
+    char *url = NULL;
+    char *referer = NULL;
+    char *user_agent = NULL;
+    struct curl_slist *headers = NULL;
 
-    /* Start up our XML-RPC client library. */
-    xmlrpc_client_init(XMLRPC_CLIENT_NO_FLAGS, NAME, VERSION);
+    char *json_str = NULL;
 
-    xmlrpc_env_init(&env);
+    response->payload = (char *) malloc(sizeof(*response->payload));
```

Comment:
This one makes no sense. payload is a pointer to char, sizeof(*response->payload) is 8 bytes on a typical 64-bit system. The buffer we have in the payload is fileld in by the response callback and there we always do reallocation.
Since you are doing `realloc()` in the response callback anyway, the response buffer can be left as it is (because we pass a zeroed buffer here) or explicitly zeroed instead here. I'd suggest to remove any payoad part handling here completely.

Reply:
True... Initial malloc has been removed.